File: queen/queen.go
package main

import (
	"flag"
	"log"
	"time"
)

var cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")

func main() {
	/*
		flag.Parse()
		if *cpuprofile != "" {
			f, err := os.Create(*cpuprofile)
			if err != nil {
				log.Fatal(err)
			}
			pprof.StartCPUProfile(f)
			defer pprof.StopCPUProfile()
		}
	*/

	queen()
	ts := time.Now()
	found := queen()
	log.Printf("found=%v, time=%v\n", found, time.Now().Sub(ts))
}

const N = 13

func check(q *[N]int, row int) int {
	if row == 0 {
		return 1
	} else {
		x0 := q[row]
		for y := 0; y < row; y++ {
			x := q[y]
			if x == x0 {
				return 0
			} else if x-x0 == row-y {
				return 0
			} else if x0-x == row-y {
				return 0
			}
		}
	}
	return 1
}

func queen() int {
	var q [N]int
	found := 0
	row := 0
	done := 0

	for done == 0 {
		if check(&q, row) > 0 {
			if row == N-1 {
				found++
			} else {
				row++
				q[row] = 0
				continue
			}
		}
		q[row] += 1
		for q[row] >= N {
			row--
			if row >= 0 {
				q[row]++
			} else {
				done = 1
				break
			}
		}
	}
	return found
}

File: queen/queen.rs
use std::time::{SystemTime, UNIX_EPOCH};

const N:i32 = 13;

fn clock_realtime() -> i64 {
    let start = SystemTime::now();
    let since_the_epoch = start.duration_since(UNIX_EPOCH).expect("Time went backwards");
    return since_the_epoch.as_millis() as i64;
}

fn array_check(array: &[i32], row: i32) -> bool 
{
    if row == 0 {
        return true;
    }
    else {
        let x0 = array[row as usize];
        for y in 0..row {
            let x = array[y as usize];
            if x == x0 {
                return false;
            }
            else if x - x0 == row - y {
                return false;
            }
            else if x0 - x == row - y {
                return false;
            }
        }
        return true;
    }
}


fn queen() -> i32 {
    let mut array = [0;N as usize];
    let mut found = 0;
    let mut row = 0;
    let mut done = false;
    while ! done {
        if array_check(&array, row) {
            if row == N - 1 {
                found += 1;
            }   else {
                row += 1;
                array[row as usize] = 0;
                continue;
            }
        }
        array[row as usize] += 1;
        while array[row as usize] >= N {
            row -= 1;
            if row >= 0 {
                array[row as usize] += 1;
            }   else {
                done = true;
                break;
            }
        }
    }
    return found;
}

fn main() {
    queen();
    let ts = clock_realtime();
    let found = queen();
    let dt = clock_realtime() - ts;
    println!("found={} time={} ms", found, dt);
}



